File: bin/analyze_warp_h5.py
import h5py
import numpy as np
import os
import sys

def analyze_h5(file_path):
    if not os.path.exists(file_path):
        print(f"Error: File {file_path} not found.")
        return

    try:
        with h5py.File(file_path, 'r') as f:
            print(f"=== Analysis of {file_path} ===")
            
            # 1. Inspect Groups and Datasets
            print("\nStructure:")
            def print_attrs(name, obj):
                print(f"  {name} ({type(obj).__name__})")
                for key, val in obj.attrs.items():
                    print(f"    Attr: {key} = {val}")
            f.visititems(print_attrs)

            # 2. Analyze Trace Data (Time Series)
            # Based on structure: simulation/trace/...
            if 'simulation' in f and 'trace' in f['simulation']:
                trace = f['simulation']['trace']
                time = np.array(trace['time'])
                density = np.array(trace['rho_mean'])
                enstrophy = np.array(trace['enstrophy'])
                
                print("\nTime-Series Summary:")
                print(f"  Steps recorded: {len(time)}")
                if len(density) > 0:
                    print(f"  Initial Mean Density: {density[0]:.6f}")
                    print(f"  Final Mean Density: {density[-1]:.6f}")
                    print(f"  Density StdDev: {np.std(density):.2e}")
                if len(enstrophy) > 0:
                    print(f"  Final Enstrophy: {enstrophy[-1]:.6e}")
                
                # Check for stability issues
                if np.any(np.isnan(density)) or np.any(np.isinf(density)):
                    print("  CRITICAL: Instability (NaN/Inf) detected in density trace!")
                elif len(density) > 0 and np.abs(density[-1] - 1.0) > 0.1:
                    print(f"  WARNING: High mass drift: {density[-1] - 1.0:+.4f}")

            # 3. Experiment Results Check
            if 'experiment/results' in f:
                res = f['experiment/results']
                md = res.attrs.get('mean_density', 0.0)
                if np.isnan(md):
                    print("\n  Forensic: 'mean_density' in results attribute is NaN. Investigating trace...")

    except Exception as e:
        print(f"Failed to analyze HDF5: {e}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python analyze_warp_h5.py <path_to_h5>")
    else:
        analyze_h5(sys.argv[1])

File: crates/gororoba_cli_data/src/bin/analyze_warp_h5.rs
//! HDF5 Warp Simulation Analysis Utility.
//!
//! Inspects HDF5 structure, trace data stability (NaN/Inf check),
//! and mass drift metrics for warp field simulations.
//!
//! Migrated from bin/analyze_warp_h5.py.

use anyhow::{Result, Context};
use hdf5::File;
use ndarray::Array1;

fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.is_empty() {
        println!("Usage: analyze-warp-h5 <path_to_h5>");
        return Ok(());
    }

    let file_path = &args[0];
    let file = File::open(file_path).with_context(|| format!("Failed to open {}", file_path))?;

    println!("=== Analysis of {} ===", file_path);

    // 1. Structural inspection (Simplified: check for key groups)
    if let Ok(sim_group) = file.group("simulation")
        && let Ok(trace_group) = sim_group.group("trace") {
            println!("\nTime-Series Summary:");
            
            if let Ok(time_ds) = trace_group.dataset("time") {
                let time: Array1<f64> = time_ds.read_1d()?;
                println!("  Steps recorded: {}", time.len());
                
                if let Ok(rho_ds) = trace_group.dataset("rho_mean") {
                    let rho: Array1<f64> = rho_ds.read_1d()?;
                    if !rho.is_empty() {
                        let initial = rho[0];
                        let final_rho = rho[rho.len() - 1];
                        println!("  Initial Mean Density: {:.6}", initial);
                        println!("  Final Mean Density: {:.6}", final_rho);
                        
                        let mut nan_found = false;
                        for &v in &rho {
                            if !v.is_finite() { nan_found = true; break; }
                        }
                        if nan_found {
                            println!("  CRITICAL: Instability (NaN/Inf) detected in density trace!");
                        } else if (final_rho - 1.0).abs() > 0.1 {
                            println!("  WARNING: High mass drift: {:+.4}", final_rho - 1.0);
                        }
                    }
                }

                if let Ok(ens_ds) = trace_group.dataset("enstrophy") {
                    let ens: Array1<f64> = ens_ds.read_1d()?;
                    if !ens.is_empty() {
                        println!("  Final Enstrophy: {:.6e}", ens[ens.len() - 1]);
                    }
                }
            }
        }

    Ok(())
}
